### PR TITLE
Encapsulate FFTs

### PR DESCRIPTION
**Pull request by MaxThevenet**
This PR proposes a wrapper around FFTs (currently FFTW and cuFFT). A new namespace `AnyFFT` is created, that contains general functions, calling either FFTW or cuFFT functions, depending on the compile options. Check changes in `SpectralFieldData.cpp` to see how it is used. It is written so that
- To use the FFT, we only need `#include "AnyFFT.H"`, and the function calls are independent of the FFT backend.
- Each backend is in a separate .cpp files (I think that'll be cleaner when we add other backends).

In general, FFT plans for different libraries hold different information. For this reason, a class `FFTPlan` was created, which contains the vendor FFT plan + metadata (pointers to src and dst and direction of the transform). Hence
```c++
FFTplans = amrex::LayoutData<cufftHandle>;
FFTplans = amrex::LayoutData<fftw_plan>;
```
was replaced by
```c++
FFTplans = amrex::LayoutData<FFTPlan>;
```
This was tested on CPU locally and on Summit (3D test `Langmuir_multi_psatd` in both cases).

However, this PR doesn't address RZ yet, this could be done in a subsequent PR.

### CODE REVIEW

**Comment by MaxThevenet** [comment]
Just put the [WIP] tag until I fix the tests and constify all of this, will let you know when this is done.

**Comment by MaxThevenet** [comment]
Travis tests pass, that's ready for review!

OpenAPI tests fail, though, but hopefully the last commit should fix it.

File: Source/FieldSolver/SpectralSolver/AnyFFT.H [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#  include <cufft.h>
+#else
+#  include <fftw3.h>
+#endif
+
+#include <AMReX_LayoutData.H>
+
+namespace AnyFFT
+{
+    // First, define library-independent types (complex, FFT plan)
```
Comment (MaxThevenet) [inline Source/FieldSolver/SpectralSolver/AnyFFT.H:14]:
```suggestion
    // First, define library-dependent types (complex, FFT plan)
```

File: Source/FieldSolver/SpectralSolver/AnyFFT.H [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    using VendorFFTPlan = cufftHandle;
+#else
+#  ifdef AMREX_USE_FLOAT
+    using VendorFFTPlan = fftwf_plan;
+#  else
+    using VendorFFTPlan = fftw_plan;
+#  endif
+#endif
+
+    // Second, define library-independent types and functions
```
Comment (MaxThevenet) [inline Source/FieldSolver/SpectralSolver/AnyFFT.H:43]:
```suggestion
    // Second, define library-independent API
```

**Comment by MaxThevenet** [comment]
I don't see the reason why the macOS test fails, it passed for the penultimate commit, while the last commit only modified comments `//`. Anyway, I think all ticks are back to green here, that's ready for review. The last few commits were to address an offline review by Rémi.

**Comment by ax3l** [comment]
OSX CI addressed in PR #1061:)

File: Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp [hunk 3]
```diff
@@ -73,107 +57,25 @@ SpectralFieldData::SpectralFieldData( const amrex::BoxArray& realspace_ba,
...
-                              CUFFT_Z2D);
-#    endif
-        if ( result != CUFFT_SUCCESS ) {
-           amrex::Print() << " cufftplan2d backward failed! Error: " <<
-            cufftErrorToString(result) << "\n";
-        }
-#  endif
+        forward_plan[mfi] = AnyFFT::CreatePlan(
+            fft_size, tmpRealField[mfi].dataPtr(),
+            reinterpret_cast<AnyFFT::Complex*>( tmpSpectralField[mfi].dataPtr()),
```
[the remark this answers is not in the document]
Comment (MaxThevenet) [reply Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp:72]: I did not look into that, this conversion was done before this PR and I just propagated it.

File: Source/FieldSolver/SpectralSolver/WrapCuFFT.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+{
+    std::string cufftErrorToString (const cufftResult& err);
+
+    FFTplan CreatePlan(const amrex::IntVect& real_size, amrex::Real * const real_array,
+                       Complex * const complex_array, const direction dir)
+    {
+        FFTplan fft_plan;
+        cufftResult result;
+        if (dir == direction::R2C){
+#if (AMREX_SPACEDIM == 3)
```
[the remark this answers is not in the document]
Comment (MaxThevenet) [reply Source/FieldSolver/SpectralSolver/WrapCuFFT.cpp:13]: You're right, I'll make the dimension a function argument, that should make it clearer and make it easy in Hipace.

[the remark this answers is not in the document]
Comment (MaxThevenet) [reply Source/FieldSolver/SpectralSolver/WrapCuFFT.cpp]: done

**Comment by MaxThevenet** [comment]
Thanks @ax3l I implemented the changes, and also merged master into this branch hoping this will fix the MacOS test.

File: Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ax3l) [reply Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp:72]: Yes, looks common :)

**Pull request merged**